src/osd/SnapMapper.cc

```diff
@@ -203,6 +217,133 @@ int SnapMapper::get_snaps(
   return 0;
 }
 
+tl::expected<std::set<snapid_t>, Scrub::mapper_req_t> SnapMapper::get_snaps(
+  const hobject_t& oid) const
+{
+  ceph_assert(check(oid));
+  using mapper_req_t = Scrub::mapper_req_t;
+
+  set<string> keys;
+  keys.insert(to_object_key(oid));
+  dout(20) << fmt::format("{}: key string: {} oid:{}", __func__, keys, oid)
+	   << dendl;
+
+  map<string, bufferlist> got;
+  int r = backend.get_keys(keys, &got);
+  if (r < 0) {
+    dout(20) << __func__ << " " << oid << " got err " << r << dendl;
+    return tl::unexpected(mapper_req_t{mapper_req_t::code_t::backend_error, r});
+  }
+  if (got.empty()) {
+    dout(20) << __func__ << " " << oid << " got.empty()" << dendl;
+    return tl::unexpected(
+      mapper_req_t{mapper_req_t::code_t::not_found, -ENOENT});
+  }
+
+  object_snaps out;
+  auto bp = got.begin()->second.cbegin();
+  decode(out, bp);
+  dout(20) << __func__ << " " << oid << " " << out.snaps << dendl;
+  if (out.snaps.empty()) {
+    dout(1) << __func__ << " " << oid << " empty snapset" << dendl;
+    ceph_assert(!cct->_conf->osd_debug_verify_snaps);
+  }
+  return out.snaps;
+}
+
+std::set<std::string> SnapMapper::expected_mapping_keys(
+  const hobject_t& clone,
+  const std::set<snapid_t>& snaps) const
+{
+  std::set<std::string> keys;
+  for (auto snap : snaps) {
+    keys.insert(to_raw_key(snap, clone));
+  }
+  dout(20) << fmt::format("{}: clone:{} snaps:{} -> keys: {}",
+			  __func__,
+			  clone,
+			  snaps,
+			  keys)
+	   << dendl;
+  return keys;
+}
+
+tl::expected<std::set<snapid_t>, Scrub::mapper_req_t>
+SnapMapper::get_mapping_entries(const hobject_t& clone,
+				std::set<snapid_t> snaps) const
+{
+  // we have the clone oid and the set of snaps relevant to this clone.
+  // Let's construct all expected SNA_ entries, and fetch them.
+  // For now - only for our shard.
+  using mapper_req_t = Scrub::mapper_req_t;
+  auto mapping_keys = expected_mapping_keys(clone, snaps);
+  map<string, bufferlist> kvmap;
+  auto r = backend.get_keys(mapping_keys, &kvmap);
+  if (r < 0) {
+    dout(10) << fmt::format("{}: backend error ({}) for clone {}",
+			    __func__,
+			    r,
+			    clone)
+	     << dendl;
+    return tl::unexpected(
+      mapper_req_t{mapper_req_t::code_t::backend_error, r});
+  }
+
+  std::set<snapid_t> snaps_from_mapping;
+  for (auto& [k, v] : kvmap) {
+    dout(20) << __func__ << " " << clone << " " << k << dendl;
+    // extract the clone ID from the value fetched for an SNA mapping key
+    auto [sn, obj] = SnapMapper::from_raw(v);
+    if (obj != clone) {
+      dout(20) << __func__ << " " << clone << " " << k << " "
+	       << "doesn't match " << obj << dendl;
+      continue;
```

Comment:
This seems like it should be returning tl::unexpected with inconsistent?

Reply:
Yes. Wasn't sure whether I am missing another explanation for this happening.
Will fix.